plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,386 @@
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
+		return "", fmt.Errorf("failed to create sse subscription %d, status: %q", sub.index, resp.Status)
+	}
+
+	return sseURL + "/" + id, nil
+}
+
+// createSubscriptionAndSseClient creates a sse subscription on the server and
+// initializes a sse client to receive sse events from the server.
+func (s *CtrlXDataLayerSSE) createSubscriptionAndSseClient(sub *Subscription) (*sseclient.SseClient, error) {
+	t, err := s.tokenManager.RequestAuthToken()
+	if err != nil {
+		return nil, err
+	}
+
+	subURL, err := s.createSubscription(sub)
+	if err != nil {
+		return nil, err
+	}
+
+	client := sseclient.NewSseClient(subURL, t.String(), s.InsecureSkipVerify)
+
+	return client, nil
+}
+
+// addMetric writes sse metric into accumulator.
+func (s *CtrlXDataLayerSSE) addMetric(se *sseclient.SseEvent, sub *Subscription) {
+	switch se.Event {
+	case "update":
+		// Received an updated value, that we translate into a metric
+		var d sseEventData
+
+		if err := json.Unmarshal([]byte(se.Data), &d); err != nil {
+			s.acc.AddError(fmt.Errorf("received malformed data from 'update' event"))
+			return
+		}
+		m, err := s.createMetric(&d, sub)
+		if err != nil {
+			s.acc.AddError(fmt.Errorf("failed to create metrics: %v", m))
+			return
+		}
+
+		s.acc.AddMetric(m)
+
+	case "error":
+		// Received an error event, that we report to the accumulator
+		var e sseEventError
+		if err := json.Unmarshal([]byte(se.Data), &e); err != nil {
+			s.acc.AddError(fmt.Errorf("received malformed data from 'error' event"))
+			return
+		}
+		s.acc.AddError(fmt.Errorf("received 'error' event for node: %q", e.Instance))
+
+	case "keepalive":
+		// Keepalive events are ignored for the moment
+		s.Log.Debug("Received keepalive event")
+
+	default:
+		// Received a yet unsupported event type
+		s.Log.Debug("Received unsupported event: %q", se.Event)
+	}
+}
+
+// createMetric - create metric depending on flag 'output_json' and data type
+func (s *CtrlXDataLayerSSE) createMetric(em *sseEventData, sub *Subscription) (telegraf.Metric, error) {
+	t := convertTimestamp2UnixTime(em.Timestamp)
+	node := sub.node(em.Node)
+	if node == nil {
+		return nil, errors.New("Node not found")
+	}
+
+	// default tags
+	tags := map[string]string{
+		"node":       em.Node,
+		"ctrlx_host": s.Hostname,
+	}
+
+	// add tags of subscription if user has defined
+	for key, value := range sub.Tags {
+		tags[key] = value
+	}
+
+	// set measurement of subscription
+	measurement := sub.Measurement
+
+	// set default field name with "value_{type}"
+	fieldName := "value_" + em.Type
+	if node.FieldName != "" {
+		// set user defined field name of node
+		fieldName = node.FieldName
+	}
+
+	if sub.OutputJSONString {
+		b, err := json.Marshal(em.Value)
+		if err != nil {
+			return nil, err
+		}
+		fields := map[string]interface{}{fieldName: string(b)}
+		m := metric.New(measurement, tags, fields, t)
+		return m, nil
+	}
```

Comment:
I don't think this is a good way of handling fields. In string form this data is not usable as-is and _requires_ further processing. Why not flatten complex types and export the values as separate fields. You can still keep arrays as comma-separated strings...

I'm making this point as you point people to the startlark processor in your `sample.conf`. However, this processor uses an interpreted language and the throughput might be limited potentially disrupting a users use-case...

Reply:
The default value of configuration `output_json_string` is false, so all complex types are flattened by default.
In some use cases we need to transform the complex data types in a processor plug-in like starlark.

Reply:
So you mean there are types that do not provide meaningful values on their own? What prevents you from processing the flattened fields? In any case you should provide a comment in the code.